### PR TITLE
added schedueled deleting of jobs if they do not exist

### PR DESCRIPTION
**Pull request by mostafa-221**
Had to change the fetch type to eager to make it work. I am not sure but we want to avoid an eager fetch type for many-to-many relationships, right?

### CODE REVIEW

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java [hunk 1]
```diff
@@ -73,4 +75,20 @@ public void scrape() throws IOException {
         log.info("Finished scraping");
         allVacancies.clear();
     }
+
+    @Scheduled(cron = "0 0 * * * *") // at second 0, minute 0 run this function
+    public void deleteNonExistingVacancies(){
+        log.info("Started deleting non-existing jobs");
+        List<Vacancy> allVacancies = vacancyService.getAllVacancies();
+        List<Vacancy> vacanciesToDelete = allVacancies.stream()
+                .filter(vacancy -> !vacancy.checkURL()) //if the url is not good anymore add it in the vacanciesToDelete
```
Comment (tkoen93) [inline src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java]: `vacancy.checkURL()` will not work for Yacht vacancies that do not exist anymore.
An example url for a removed Yacht vacancy is https://www.yacht.nl/vacatures/9079222/data-analist---modeler which returns status 200.

Quick research shows me that if you use a random vacancy url for HuxleyIT it returns status 410. Jobbird returns status 301.

I might fix this myself this week as you should be enjoying your time :-)

**Comment by tkoen93** [comment]
Noticed that you want to merge directly into master. Might be better to first merge into develop where we can collect all our mvp fixes/features. When everything works as it should we can easily merge develop into master.

I believe I can change that myself, so will do after finishing this comment.

Edit: getting some warning, might need to start a new merge request?

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java (hunk 1 above)
Comment (tkoen93) [reply src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java]: Resolved https://github.com/mostafa-221/jobcrawler-backend/pull/19/commits/ff26b0d0770003ab93ead9abb80fd9c414483941

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java [hunk 2]
```diff
@@ -73,4 +69,25 @@ public void scrape() throws IOException {
...
         allVacancies.clear();
     }
+
+    @Scheduled(cron = "0 30 11,17 * * *") // Runs two times a day. At 11.30am and 5.30pm.
+    public void deleteNonExistingVacancies() {
+        log.info("CRON Scheduled -- Started deleting non-existing jobs");
+        List<Vacancy> allVacancies = vacancyService.getAllVacancies();
+        List<Vacancy> vacanciesToDelete = allVacancies.stream()
+                .filter(vacancy -> {
+                    if (vacancy.getBroker().equals("Yacht"))
```
Comment (ghyze) [inline src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java]: It's best practice to switch this equals statement around.
Use `"Yacht".equals(vacancy.getBroker())`.
The reasoning behind this is that you if `vacancy.getBroker()` returns null, this code will still work.

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java [hunk 3]
```diff
@@ -73,4 +69,25 @@ public void scrape() throws IOException {
...
+                        return !yachtVacancyScraper.doesVacancyExist(vacancy.getVacancyURL());
```
Comment (ghyze) [inline src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java:80]: Although not strictly necessary, using curly braces will make the code more robust. Not everyone will notice the lack of curly braces, and some developers will assume that they are in a code block, which would result in unexpected behavior.

File: src/main/java/nl/ordina/jobcrawler/service/VacancyService.java [hunk 4]
```diff
@@ -32,8 +33,12 @@ public Vacancy add(Vacancy vacancy) {
         vacancy.setSkills(existingSkills); // add the skills to the vacancy
 
 
-        vacancy.checkURL(); //checking the url, if it is malformed it will throw a VacancyURLMalformedException
-        return vacancyRepository.saveAndFlush(vacancy); // save and send to the database
+        if (vacancy.checkURL()){    //checking the url, if it is malformed it will throw a VacancyURLMalformedException
+            return vacancyRepository.saveAndFlush(vacancy);
+        } else {
+            throw new VacancyURLMalformedException("Website could not be reached");
+        }
```
Comment (ghyze) [inline src/main/java/nl/ordina/jobcrawler/service/VacancyService.java]: This looks odd. What is happening here exactly? And what is supposed to happen?

**Review by ghyze — DISMISSED** [review]

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java (hunk 2 above)
Comment (mostafa-221) [reply src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java]: done

File: src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java (hunk 3 above)
Comment (mostafa-221) [reply src/main/java/nl/ordina/jobcrawler/service/VacancyStarter.java:80]: done

File: src/main/java/nl/ordina/jobcrawler/service/VacancyService.java (hunk 4 above)
Comment (mostafa-221) [reply src/main/java/nl/ordina/jobcrawler/service/VacancyService.java]: the name was changed to hasValidURL. So if a vacancy has a valid URL it saves it, else it throws an exception

**Review by tkoen93 — APPROVED** [review]

**Pull request merged**